In [1]:
!pip install pandas seaborn matplotlib scikit-learn


[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import pandas as pd
df = pd.read_csv('preprocessed_data.csv')

df.head()

Make                   Model  Year     Price  Mileage  \
0  toyota                   camry  2016   47819.0   156500   
1     kia                 sorento  2013   61250.0   169543   
2    mini                  cooper  2023   31861.0   221583   
3  nissan                  altima  2016  110322.0    69754   
4  toyota  land-cruiser-76-series  2020  139994.0    71399   

              Body Type  Cylinders            Transmission Fuel Type  Color  \
0                 Sedan          4  Automatic Transmission  Gasoline  Black   
1                   SUV          4  Automatic Transmission  Gasoline   Grey   
2  Soft Top Convertible          4  Automatic Transmission  Gasoline   Grey   
3                 Sedan          4  Automatic Transmission  Gasoline    Red   
4         Pick Up Truck          4     Manual Transmission  Gasoline  White   

   ...  Body Type_Wagon   Make Encoded  Model Encoded  Color Encoded  \
0  ...              0.0  111815.739082   59573.661538  295369.325024   
1  ...              0.0   44691.069565   39255.111111  259363.792655   
2  ...              0.0   75449.916667   72200.627907  259363.792655   
3  ...              0.0   95783.971892   46364.504505  276426.754331   
4  ...              0.0  111815.739082  137098.800000  190510.761848   

   Year zscore scaled  Mileage zscore scaled  Price zscore scaled  \
0            0.263740               0.015991            -0.419182   
1           -0.254345               0.171863            -0.390664   
2            1.472606               0.793774            -0.453067   
3            0.263740              -1.020677            -0.286467   
4            0.954521              -1.001018            -0.223462   

   Cylinders zscore scaled  Car Age zscore scaled  \
0                -1.155725              -0.263740   
1                -1.155725               0.254345   
2                -1.155725              -1.472606   
3                -1.155725              -0.263740   
4                -1.155725              -0.954521   

   Mileage Per Year zscore scaled  
0                       -0.252931  
1                       -0.331979  
2                        2.011678  
3                       -0.486622  
4                       -0.328313  

[5 rows x 40 columns]

In [6]:
X = df.drop(columns=['Price', 'Year', 'Mileage', 'Cylinders', 'Log Price', 'Fuel Type', 'Make', 'Model', 'Body Type', 'Transmission', 'Color', 'Price zscore scaled'])  
y = df['Price'] 

In [7]:
X

Car Age  Mileage Per Year  Transmission Encoded  Fuel Type_Diesel  \
0           9      17388.888889                     0               0.0   
1          12      14128.583333                     0               0.0   
2           2     110791.500000                     0               0.0   
3           9       7750.444444                     0               0.0   
4           5      14279.800000                     1               0.0   
...       ...               ...                   ...               ...   
9995        7      10988.571429                     0               0.0   
9996        3      86050.000000                     0               0.0   
9997       11       7320.454545                     0               0.0   
9998        9      32033.888889                     0               1.0   
9999       16      17728.000000                     0               0.0   

      Fuel Type_Electric  Fuel Type_Gasoline  Fuel Type_Hybrid  \
0                    0.0                 1.0               0.0   
1                    0.0                 1.0               0.0   
2                    0.0                 1.0               0.0   
3                    0.0                 1.0               0.0   
4                    0.0                 1.0               0.0   
...                  ...                 ...               ...   
9995                 1.0                 0.0               0.0   
9996                 0.0                 1.0               0.0   
9997                 0.0                 1.0               0.0   
9998                 0.0                 0.0               0.0   
9999                 0.0                 1.0               0.0   

      Body Type_Coupe  Body Type_Crossover  Body Type_Hard Top Convertible  \
0                 0.0                  0.0                             0.0   
1                 0.0                  0.0                             0.0   
2                 0.0                  0.0                             0.0   
3                 0.0                  0.0                             0.0   
4                 0.0                  0.0                             0.0   
...               ...                  ...                             ...   
9995              0.0                  0.0                             0.0   
9996              0.0                  0.0                             0.0   
9997              0.0                  0.0                             0.0   
9998              0.0                  0.0                             0.0   
9999              0.0                  0.0                             0.0   

      ...  Body Type_Van  Body Type_Wagon   Make Encoded  Model Encoded  \
0     ...            0.0              0.0  111815.739082   59573.661538   
1     ...            0.0              0.0   44691.069565   39255.111111   
2     ...            0.0              0.0   75449.916667   72200.627907   
3     ...            0.0              0.0   95783.971892   46364.504505   
4     ...            0.0              0.0  111815.739082  137098.800000   
...   ...            ...              ...            ...            ...   
9995  ...            0.0              0.0  249065.056604  224851.305556   
9996  ...            0.0              0.0  190098.483627   58581.117647   
9997  ...            0.0              0.0  111815.739082  104034.322917   
9998  ...            0.0              0.0   42722.228571   39484.200000   
9999  ...            0.0              0.0  372082.499327  141566.256410   

      Color Encoded  Year zscore scaled  Mileage zscore scaled  \
0     295369.325024            0.263740               0.015991   
1     259363.792655           -0.254345               0.171863   
2     259363.792655            1.472606               0.793774   
3     276426.754331            0.263740              -1.020677   
4     190510.761848            0.954521              -1.001018   
...             ...                 ...                    ...   
9995  190510

In [8]:
# split dataset into train and test sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [9]:
# Linear Regression

from sklearn.linear_model import LinearRegression

# Initialize model
lr_model = LinearRegression()

# Train model
lr_model.fit(X_train, y_train)

# Predict on test set
y_pred = lr_model.predict(X_test)

In [11]:
# Evaluate Linear Regression Model

from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, root_mean_squared_error

# Linear Regression
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse =root_mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Linear Regression - MAE: {mae}, MSE: {mse}, RMSE: {rmse}, R²: {r2}")

Linear Regression - MAE: 102452.46093634289, MSE: 39626365872.926155, RMSE: 199063.72314644916, R²: 0.8488557828043372


R² value is 0.85, which suggests that the model explains about 85% of the variance in the target variable (used vehicle prices)

Since Mean Squared Error (MSE) is sensitive to large errors, therfore having large value like 39,626,365,872.93 for MSE means that there are significant outliers in the selected features. So, we might have to analyze the selected feature further and handle those outliers to get better accuracy.